### ISSUE TITLE
GDExtension - get_instance_id seems to return invalid IDs 

### ISSUE THREAD

**Issue by SeleDreams** [godotengine/godot-cpp#1061]
I'm posting this as i'm having some issues porting a godot 3 project because of it
basically, to guarantee I am only holding valid references, I use instance ids to make sure the references are valid
however it seems like currently "UtilityFunctions::is_instance_id_valid" will always return false for the id returned by get_instance_id of a node even for valid node references. And I checked, the ID itself seems to be wrong as calling "UtilityFunctions::instance_from_id" causes a crash as well

**Comment by SeleDreams** [comment]
I am on godot 4 stable and on the godot 4 stable branch of godot-cpp as well

**Comment by follower** [comment]
This may be entirely unrelated but I happened to remember seeing this recent commit, so thought I'd mention it in case it's related/useful:

 * https://github.com/godotengine/godot/commit/c5f7a89b0f557c9540fa2c6af2cbb3a0fb5096a3 (a.k.a. https://github.com/godotengine/godot/pull/73892)

**Comment by SeleDreams** [comment]
I don't think this is related

**Comment by SeleDreams** [comment]
![image](https://user-images.githubusercontent.com/16335601/222814421-c526926f-0cd6-45c0-8ede-5a1c38fe1161.png)
I think the issue in general might be bigger since even when I am getting valid nodes from godot-cpp in fact calling functions from them returns read access violations

**Comment by dsnopek** [comment]
> I think the issue in general might be bigger since even when I am getting valid nodes from godot-cpp in fact calling functions from them returns read access violations

The access violation could be a manifestation of the `RayCast3D` not actually being a `RayCast3D` on the godot-cpp side: https://github.com/godotengine/godot-cpp/issues/995

To summarize that issue: currently, if you pass a Godot object to a function defined in GDExtension that takes a class higher up in the class tree (say it takes `Node3D` and you passed in a `RayCast3D`), it'll actually get created as the higher up class in godot-cpp (so, in my example, it'd be a `Node3D` on the godot-cpp side, even though it was a `RayCast3D` on the Godot side).

This PR attempts to fix that issue: https://github.com/godotengine/godot-cpp/pull/1050

If you felt adventurous you could test the PR and see if it helps? (It requires a change in Godot as well as godot-cpp)

I don't know if it'd do anything for the original `get_instance_id()` issue in the description, but it may help with the issue screenshotted in your last comment

**Comment by SeleDreams** [comment]
> > I think the issue in general might be bigger since even when I am getting valid nodes from godot-cpp in fact calling functions from them returns read access violations
> 
> The access violation could be a manifestation of the `RayCast3D` not actually being a `RayCast3D` on the godot-cpp side: #995
> 
> To summarize that issue: currently, if you pass a Godot object to a function defined in GDExtension that takes a class higher up in the class tree (say it takes `Node3D` and you passed in a `RayCast3D`), it'll actually get created as the higher up class in godot-cpp (so, in my example, it'd be a `Node3D` on the godot-cpp side, even though it was a `RayCast3D` on the Godot side).
> 
> This PR attempts to fix that issue: #1050
> 
> If you felt adventurous you could test the PR and see if it helps? (It requires a change in Godot as well as godot-cpp)
> 
> I don't know if it'd do anything for the original `get_instance_id()` issue in the description, but it may help with the issue screenshotted in your last comment

thanks for the suggestion, i'm gonna try the branch from your PR

**Comment by SeleDreams** [comment]
So, I tried your PR and it did change my issue, but it doesn't seem like it fully fixed it as it changed where the issue is 
![image](https://user-images.githubusercontent.com/16335601/222819931-2bd64f0a-4a89-47fa-bae7-8bf84b7fd5ec.png)

**Comment by dsnopek** [comment]
Ah, well, it was worth a try :-)

**Comment by SeleDreams** [comment]
is there a workaround then ? since if i can't use object::cast_to none of my nodes can get their children

**Comment by Manonox** [comment]
Bumping this, because got something very similar

Segfaults when I try to get something with UtilityFunctions::instance_from_id, but returned ids by get_instance_id are valid

This is GDScript that I am testing this with
```js
func _ready():
    var id = get_instance_id()
    print(id)
    print(instance_from_id(id))
    print(funnyfunction(id))
```

C++ function:
```cpp
bool TestThingy::funnyfunction(int64_t instance_id) {
    int64_t gde_instance_id = get_instance_id();
    UtilityFunctions::print(gde_instance_id);
    
    Object *object = UtilityFunctions::instance_from_id(instance_id); // segfault
    if (object != nullptr)
        UtilityFunctions::print(object->to_string());
    return instance_id == gde_instance_id;
}
```

Output:
```
26071794816
TestThingy:[Wrapped:0]
26071794816
// Aaaand it segfaults here
```

**Comment by prothegee** [comment]
Im not sure that the goal is to store data for custom c++ class or not,

but if to store data class, I prefer to use string to locate correct child node then store it on *preferred root class node*

Says we want to store `CanvasLayer` or custom class that inherent by that *class* to some root node & access it later

---

<details>
<summary>code</summary>
<br>

`UiCanvasLayer.h`
```c++
// custom ui canvas layer from godot CanvasLayer
class CUiCanvasLayer final : public godot::CanvasLayer
{
    GDCLASS(CUiCanvasLayer, godot::CanvasLayer);


private:
    /* data */


protected:
    static void _bind_methods() {};


public:
    CUiCanvasLayer(/* args */) {};
    ~CUiCanvasLayer();

    void            _enter_tree() {};
    void            _exit_tree() {};
    void            _ready() {};
    void            _process(float delta) {};

    void            SayHello()
                    {
                        godot::UtilityFunctions::print("// Hello from CUiCanvasLayer.");
                    }
};
```

---

`Child.h`
```c++
#include "UiCanvasLayer.h"


// storage for custom CanvasLayer type node
struct SChildRootCanvasLayerStorage
{
    int                             m_nodeId;
    std::string                     m_nodeName;
    std::shared_ptr<CUiCanvasLayer> m_pNode;
    godot::CanvasLayer              *m_pCanvasLayer;
};
```

---

`NodeRoot.h`
```c++
#include "Child.h"

struct SChildRootCanvasLayerStorage;

// custom node from Node godot class
class CNodeRoot final : public godot::Node
{
    GDCLASS(CNodeRoot, godot::Node);


private:
    std::vector<std::pair<std::string, SChildRootCanvasLayerStorage>>     m_childsCanvasLayerStorage;


protected:
    static void _bind_methods() {};


public:
    CUiCanvasLayer(/* args */) {};
    ~CUiCanvasLayer();

    void            _enter_tree() {};
    void            _exit_tree() {};
    void            _ready();
    void            _process(float delta) {};

public:
    std::vector<std::pair<std::string, SChildRootCanvasLayerStorage>>     GetChildsCanvasLayerStorage() const { return m_childsCanvasLayerStorage; }
};
```

---

`NodeRoot.cc`
```c++
...

void CNodeRoot::_ready()
{
    for (int i = 0; i < get_child_count(); i++)
    {
        SChildRootCanvasLayerStorage    childNode;
        godot::String                   child{get_child(i)->get_name()};
        std::string                     childString(child.utf8().get_data());
        
        // compare first from string range then assign it if correct
        if (childString.rfind("ui_canvas-", 0) == 0)
        {
            childNode.m_nodeId       = i;
            childNode.m_nodeName     = childString;
            childNode.m_pNode        = std::make_shared<CUiCanvasLayer>();
            childNode.m_pCanvasLayer = get_node<godot::CanvasLayer>(childString.c_str());

            #ifdef NDEBUG
            #else
                std::string msg = "-- DEBUG: child name of '" + childString + "' HAS 'ui_canvas-' prefix";
                godot::UtilityFunctions::print(msg.c_str());
            #endif

            m_childsCanvasLayerStorage.emplace_back(childString, childNode);
        }
    }
    #ifdef NDEBUG
    #else
        godot::UtilityFunctions::print("-- DEBUG: m_childsCanvasLayerStorage info:");
        for (auto &data : m_childsCanvasLayerStorage)
        {
            std::string baseData        = "          " + data.first + " is came from CUiCanvasLayer";

            godot::String objectName    = data.second.m_pCanvasLayer->get_name();
            std::string nativeData      = "          ";
            nativeData += objectName.utf8().get_data();
            nativeData += " is came from godot::CanvasLayer";

            godot::UtilityFunctions::print(baseData.c_str());
            godot::UtilityFunctions::print(nativeData.c_str());

            // invoke SayHello function from one of CUiCanvasLayer where the object name is 'ui_canvas-main_start' from storage object
            if (data.first == "ui_canvas-main_start")
            {
                data.second.m_pNode->SayHello();
            }
        }
    #endif
}

...
```

</details>

<br>

[preview](https://drive.google.com/file/d/1MVrZKS23vfUdsBG7XfSzv64AjNbyvZ1X/preview)
<img width="1024" height="275" src="https://drive.google.com/uc?export=view&id=1MVrZKS23vfUdsBG7XfSzv64AjNbyvZ1X">

**Comment by SeleDreams** [comment]
It is pretty inefficient to store strings, first of all it uses way more ram than necessary but also getting the node out of strings id way more costly than getting them out of their ID

**Comment by prothegee** [comment]
> It is pretty inefficient to store strings, first of all it uses way more ram than necessary but also getting the node out of strings id way more costly than getting them out of their ID

true, way better with int8_t & specific node for further implementation
